mount/mount.go

```diff
@@ -22,3 +47,48 @@ func All(mounts []Mount, target string) error {
 	}
 	return nil
 }
+
+// TempMount mounts the provided mounts to a temp dir, and pass the temp dir to f.
+// The mounts are valid during the call to the f.
+// Finally we will unmount and remove the temp dir regardless of the result of f.
+func TempMount(mounts []Mount, m Mounter, f func(root string) error) error {
+	if len(mounts) == 0 {
+		return errdefs.ErrInvalidArgument
+	}
+	if m == nil {
+		m = mounter{}
+	}
+
+	root, err := ioutil.TempDir("", "containerd-TempMount")
+	if err != nil {
+		return errors.Wrapf(err, "failed to create %s", root)
+	}
+	// We use Remove here instead of RemoveAll.
+	// The RemoveAll will delete the temp dir and all children it contains.
+	// When the Unmount fails, if we use RemoveAll, We will incorrectly delete data from mounted dir.
+	// if we use Remove,even though we won't successfully delete the temp dir,
+	// but we only leak a temp dir, we don't loss data from mounted dir.
+	// For details, please refer to #1868 #1785.
+	defer func() {
+		if err := os.Remove(root); err != nil {
+			log.L.Errorf("Failed to remove the temp dir %s: %v", root, err)
+		}
+	}()
+
+	// We should do defer first, or we will not do Unmount when only partful Mount fail.
+	defer func() {
+		if uerr := m.Unmount(root); uerr != nil {
+			uerr = errors.Wrapf(uerr, "failed to unmount %s", root)
+			if err == nil {
+				err = uerr
```

Comment:
This doesn't have the same behaviour anymore because `err` is not a named return.  You'll need to change the return value to `(err error)`.

I often find checking named returned in a defer is not very obvious, but at least with this change it's only in a single place, instead of 7.

A unit test for this might be good, since there are a bunch of error cases. If the `m Mounter` arg is removed unit testing the errors may be more difficult, but a perhaps a shim could be used for mount/unmount to make it easier.

Reply:
You are correct. I missed the named return. I has been added it.
For unit test I have no good idea now. I will consider it.